### ISSUE TITLE
Accessibility: add attribute for aria-describedby for validate purpose

### ISSUE THREAD

**Issue by hamsterbacke81** [primefaces/primefaces#13896]
Hi, 

this case is rather complicated. I hope I can explain this properly.

**first case**:
code from [showcase](http://www.primefaces.org/showcase/ui/input/inputText.xhtml?jfwid=19660)
sorry, I don't know how to properly format the code

```xml
<div class="field">
        <p:outputLabel for="username2" styleClass="block">Username</p:outputLabel>
        <p:inputText id="username2" styleClass="block ui-state-error"/>
        <small class="block ui-state-error">Username is not available.</small>
    </div>
```

![Image](https://github.com/user-attachments/assets/d9679c71-eee2-4a3e-9864-b4e9bfac21bc)

this create a simple input field with a validation error message. However, when using keyboard for navigation and tab into the field, the error message will not be read out. This can be achieved by adding `aria-describedby` to the input, and since `aria-describedby` is not supported in inputText, I have to use passthrough:

```xml
<div class="field">
        <p:outputLabel for="username2" styleClass="block">Username</p:outputLabel>
        <p:inputText id="username2" styleClass="block ui-state-error" pt:aria-describedby="errorUsername"/>
        <small id="errorUsername" class="block ui-state-error">Username is not available.</small>
    </div>
```

this works fine. However, with this only usable with passthrough, more problems occur:

**second case**
When the input field is in a form (which they mostly are), referencing to a id become more complicated:

```xml
<h:form id="personalForm">
        <div class="field">
        <p:outputLabel for="username2" styleClass="block">Username</p:outputLabel>
        <p:inputText id="username2" styleClass="block ui-state-error" pt:aria-describedby="errorUsername"/>
        <small id="errorUsername" class="block ui-state-error">Username is not available.</small>
    </div>
    </h:form>
```

thanks to the auto-generated id, `pt:aria-describedby="errorUsername` won't work anymore, you need `pt:aria-describedby="personalForm:errorUsername`. And when form get more complex, the id will get longer.

**third case**
In my case, I have a form in which there is a list of dynamically generated input fields, for this I use <ui:repeat id="names">. In the generated HTML output, the error message container will get the `id="personalForm:names:0:errorUsername"`. So now, I can't reference to this id at all, right? Since pt only pass the whole thing through as string, something like `pt:aria-describedby="personalForm:names:{index}:errorUsername"` will not work.

```xml
<h:form id="personalForm">
        <div class="form-fields">
            <ui:repeat value="#{testView.list}" var="item" id="names">
                <div class="field-container">
                    <span class="ui-float-label">
                        <p:inputText value="#{item.name}"
                                     pt:aria-describedby="personalForm:names:{index}:errorUsername">
                        </p:inputText>
                        <p:outputLabel for="@previous" value="Name" />
                    </span>
                </div>
                <p:fragment id="errorUsername">
                    <small class="block ui-state-error">Username is not available.</small>
                </p:fragment>
            </ui:repeat>
        </div>
    </h:form>
```

So, the "most simple" way will be add a new attribute for `aria-describedby` (I think in primevue this exists already?). I know it is probably not that simple, because inputText is only one example, it should exits in all form control. 
In the meantime, is there any other way to reference to a generated id in a <ui:repeat>?

Thanks for your work.

**Comment by melloware** [comment]
@tandraschko there are two things here.

1. I agree i think `ariaDescribedBy` should be on all `input` fields.

2. I tried this which uses `@next` so you don't have the name the component and makes the `<span>` a JSF component by using `faces:id`

```xml
<p:outputLabel for="@next" styleClass="block">Username</p:outputLabel>
<p:inputText id="username2" styleClass="block ui-state-error" pt:aria-describedby="@next" />
<small faces:id="errorUsername" class="block ui-state-error">Username is not available.</small>
```

the `@next` working in the p:outputLabel `for` but doesn't work in the `pt:`.  I wonder if we should run `pt: elements through the same processing we do for `for`?

it makes it this...

```xml
<label id="frmInput:j_idt155" class="ui-outputlabel ui-widget block" for="frmInput:username2">Username</label>
<input id="frmInput:username2" name="frmInput:username2" 
    type="text" class="ui-inputfield ui-inputtext ui-widget ui-state-default block ui-state-error" aria-describedby="@next">
<small id="frmInput:errorUsername" class="block ui-state-error">Username is not available.</small>
```

**Comment by hamsterbacke81** [comment]
@melloware 
`@next` will only work for the adjasent element as far as I understand? So if I use, i.e. ui-float-label, which normally renders to 
```xml
<span class="ui-float-label">
    <input id="someId"...>
    <label class="ui-outputlabel ui-widget">Name</label>
<span>
<small id="generatedIdErrorUsername">
```

or, if the error message container is not direckly after the input, it won't work anymore. I had something similar which I had to use `<p:outputLabel for="@form:inputXYZ">` because I had a tooltip between the label and the input field.

**Comment by melloware** [comment]
yes `@next` and `@previous` look for the next or previous JSF component in the tree

**Comment by hamsterbacke81** [comment]
in the meantime, is there any other workarround to reference to the generated id in a `<ui:repeat>`. Something like `pt:aria-describedby="personalForm:names:{index}:errorUsername"` will probably not work since pt only pass the whole thing through as a string.

**Comment by melloware** [comment]
yeah that won't work...

**Comment by melloware** [comment]
with my addition now...

```xml
<h:form id="frmTest">
	<div class="field">
		<p:outputLabel for="username2"
		               styleClass="block">Username</p:outputLabel>
		<p:inputText id="username2"
		             styleClass="block ui-state-error"
		             ariaDescribedBy="errorUsername"/>
		<small faces:id="errorUsername"
		       class="block ui-state-error">Username is not available.</small>
	</div>
</h:form>
```

now gets translation to...

```xml
<div class="field">
	<label id="frmTest:j_idt9" class="ui-outputlabel ui-widget block" for="frmTest:username2">Username</label>
	
        <input id="frmTest:username2" name="frmTest:username2" type="text" 
class="ui-inputfield ui-inputtext ui-widget ui-state-default block ui-state-error" 
aria-describedby="frmTest:errorUsername">
	
         <small id="frmTest:errorUsername" class="block ui-state-error">Username is not available.</small>
</div>		
```

**Comment by melloware** [comment]
In theory my solution should work inside a UI Repeat as well.

**Comment by hamsterbacke81** [comment]
this will come with 16.0.0?

**Comment by melloware** [comment]
yes slated for 16.0.0

**Comment by tandraschko** [comment]
this change destroys my applications:

```
2025-11-19 15:56:34 SEVERE (executor-thread-11) Cannot find component for expression "form:crud:form:groups:0:fields:1:j_id_4v" referenced from "form:crud:form:groups:0:fields:1:inputSelectOne".: jakarta.faces.component.search.ComponentNotFoundException: Cannot find component for expression "form:crud:form:groups:0:fields:1:j_id_4v" referenced from "form:crud:form:groups:0:fields:1:inputSelectOne".
	at org.apache.myfaces.component.search.SearchExpressionHandlerImpl.resolveComponent(SearchExpressionHandlerImpl.java:255)
	at org.primefaces.expression.SearchExpressionUtils.contextlessResolveComponent(SearchExpressionUtils.java:91)
	at org.primefaces.expression.SearchExpressionUtils.contextlessResolveComponent(SearchExpressionUtils.java:79)
	at org.primefaces.renderkit.InputRenderer.renderAccessibilityAttributes(InputRenderer.java:149)
	at org.primefaces.renderkit.InputRenderer.renderAccessibilityAttributes(InputRenderer.java:117)
```

and this is white expected with the code changes.

If you check OutputLabelRenderer, the labelledBy takes a absolute clientId, not a expression. If you now try to resolve expresions, it wont work and is a massive performance overhead.

**Comment by tandraschko** [comment]
the case is quite easy:

```
                                    <p:outputLabel for="@next" value="#{formController.getLabel(field, cc.attrs.data)}" />
                                    <p:selectOneMenu id="inputSelectOne" value="#{formController.wrapValue(field, cc.attrs.data, component, cc.attrs.update).value}"
                                                     disabled="#{formController.isReadonly(field, cc.attrs.data) or cc.attrs.readonly}"
                                                     required="#{formController.isRequired(field, cc.attrs.data)}"
                                                     converter="#{field.getConverter().get()}">
                                        <f:selectItems value="#{formController.getSelectItems(field, cc.attrs.data)}" />
                                        <p:ajax event="change" update="#{formController.resolveUpdates(field, cc.attrs.update)}" process="@this"/>
                                    </p:selectOneMenu>
```

**Comment by melloware** [comment]
I think i see the issue

**Comment by melloware** [comment]
PR submitted

**Comment by tandraschko** [comment]
we actually got more issues:

1) is currently labelledBy only usable internal or also by users?
2) ariaDescripedBy actually allows multiple elements - or code only allows 1

**Comment by melloware** [comment]
1. i actually don't think labelledBy is used by Users or they would have seen the issue before i think its internal only.
2. `ariaDescribedBy` should only contain 1 ID that it points to the DOM element that describes it.  It should not use many elements there.  In the uses cases above we need to be able to use `@next` `@previous` or an `EL epxression` as it could evaluate to many different things similar to `p:outputLabel`

**Comment by melloware** [comment]
I wonder if it has to walk the Composite components like `outputLabel` does?

**Comment by tandraschko** [comment]
1) good, then we should not touch that at all
2) the HTML spec shows it can be used with multiple elements, so SearchExpressionUtils.resolveClientIds can be used.
3) about CC: i dont know... CC is complex

**Comment by melloware** [comment]
ok let me update!  But my current PR should fix your immediate issue
